### DISCUSSION TITLE
Upgrading Nextjs version from 9.5.2 to 12.1.6, causing CircleCI build to faild with error Code 137

### DISCUSSION THREAD

**Discussion by the-mgi** [vercel/next.js#36982] [Help]
This is my package.json file
```{
  "name": "with-redux",
  "version": "1.0.0",
  "scripts": {
    "build": "NODE_OPTIONS='--max-old-space-size=1536' next build",
    "start": "next start",
    "start-dev": "next start -p 5001",
    "start-production": "next start -p 5002",
    "dev": "next dev -p 3202"
  },
  "dependencies": {
    "@ant-design/icons": "^4.7.0",
    "@emotion/styled": "11.8.1",
    "@material-ui/core": "^4.12.4",
    "@material-ui/icons": "^4.11.3",
    "@material-ui/lab": "^4.0.0-alpha.61",
    "@webdeb/next-styles": "^1.1.1",
    "@zeit/next-css": "^1.0.1",
    "antd": "^4.20.5",
    "aws-s3": "^2.0.5",
    "aws-sdk": "^2.1136.0",
    "axios": "^0.27.2",
    "axios-curlirize": "^2.0.0",
    "babel-plugin-import": "^1.13.5",
    "card-validator": "^8.1.1",
    "classnames": "^2.3.1",
    "cookie": "^0.5.0",
    "formik": "^2.2.9",
    "framer-motion": "^6.3.3",
    "immutability-helper": "^3.1.1",
    "js-cookie": "^3.0.1",
    "less": "^3.13.1",
    "less-loader": "^10.2.0",
    "less-vars-to-js": "^1.3.0",
    "lodash": "^4.17.21",
    "moment": "^2.29.3",
    "moment-timezone": "^0.5.34",
    "next": "12.1.6",
    "next-cookies": "^2.0.3",
    "next-seo": "^5.4.0",
    "next-with-less": "^2.0.5",
    "notistack": "1.0.10",
    "nprogress": "^0.2.0",
    "null-loader": "^4.0.1",
    "react": "^17.0.1",
    "react-alice-carousel": "^2.6.1",
    "react-dom": "^17.0.1",
    "react-hook-form": "7.31.1",
    "react-image": "^4.0.3",
    "react-lazy-load-image-component": "^1.5.4",
    "react-recaptcha": "^2.3.10",
    "react-redux": "^8.0.1",
    "react-responsive-carousel": "^3.2.23",
    "react-select": "^5.3.2",
    "redux": "^4.2.0",
    "redux-devtools-extension": "^2.13.9",
    "redux-thunk": "^2.4.1",
    "sass": "^1.51.0",
    "styled-components": "^5.3.5",
    "styled-jsx": "^5.0.2",
    "swr": "^1.3.0",
    "yup": "^0.32.11"
  },
  "devDependencies": {
    "babel-plugin-styled-components": "2.0.7"
  }
}
```

[Linked Issue](https://github.com/vercel/next.js/issues/7929)

## this is .babelrc file
```
{
  "presets": [
    "next/babel"
  ],
  "plugins": [
    [
      "styled-components",
      {
        "ssr": true,
        "displayName": true,
        "preprocess": false
      }
    ]
  ]
}
```

also couldn't opt in for the swc compiler, as i could not find the replacement for the babel plugin.

## this is the next.config.js
```
const withLess = require("next-with-less");
module.exports = withLess({
  // reactStrictMode: true,
  lessLoaderOptions: {},
  publicRuntimeConfig: {
    DOMAIN_BASE_URL: process.env.DOMAIN_BASE_URL,
    SERVICE_API_BASE_URL: process.env.SERVICE_API_BASE_URL,
    GOOGLE_TAG: process.env.GOOGLE_TAG,
    API_BASE_URL:process.env.API_BASE_URL
  },
});
```

Local Machine Spec

- ubuntu 22.04
- ram 24 gbs
- node version: 14.17.3
- npm version: 8.10.0

Server Specs

- ubuntu 18.04
- ram 2 gb
- node version: 14.16.1
- npm version: 6.14.12

What i have tried
1. tried setting NODE_OPTIONS=--max-old-space-size to 2048 of no use
2. tried to build on my local machine, eveyrthing seems to be working fine.
3. tried on a separate docker container on my local machine, copying the app to container and then tried building, everything works fine there as well.
4. this issue is just something with the ec2 instance or nextjs stuff i am not sure.

**Comment by icyJoseph** [comment]
A few more things to try out:

- Downgrade NPM locally to the same version (6.14.12), and build again
- Upgrade to the latest next 11 (11.1.3)
- Which plugin couldn't you find a replacement for?

**Reply by the-mgi** [reply]
yeah so, i downgraded npm to 6.14.12 on my local with nextjs version 11.1.1, unfortunately the build is successful on my local machine, but not on the server i.e ec2 instance t3.small with 2gb of ram.

and for the plugin ```"babel-plugin-styled-components"``` i couldn't find any replacement.

can you suggest me some other stuff to try out for the build to be successful on the server. that would be of great help, been tackling this issue for the last week, up till now with no luck.
thanks! :)

**Reply by icyJoseph** [reply]
You just need to do, like it's shown here, https://github.com/vercel/next.js/blob/canary/examples/with-styled-components/next.config.js

```js
/** @type {import('next').NextConfig} */

const nextConfig = {
  reactStrictMode: true,
  compiler: {
    styledComponents: true,
  },
}

module.exports = nextConfig
```

And I think you can drop the entire babel plugin. That's for next 12.1.x though.

**Reply by the-mgi** [reply]
cool, got it for the babel plugin.

anything for the build?

**Reply by icyJoseph** [reply]
I see, so Next 11.1.1 also failed.

Perhaps this is a webpack issue?

I'd also try to remove the next-with-less plugin, just for testing. 

There's also this package, babel-plugin-import, which you do not seem to use? Remove it if so.

Also, if you have time, checkout a branch, and deploy as little as possible of your application, like, just one page, and its dependencies, that should work.

I also found this issue: https://github.com/elado/next-with-less/issues/3, not sure it applies here.

Also, why not increase the ec2 memory, if doubling works, then ok... but if that doesn't work, then it might be that some process is just looping through accumulating objects.

**Reply by the-mgi** [reply]
yeah so, i deleted the files and added them gradually {not all}, the build succeed but as soon as i added more files, and folders in the pages folder the build starts failing.

from this i guess we can infer that this is not a problem a next-with-less plugin.

the last resort is to see if the increase ec2 memory helps, If that bails on me, i don't know what to do man, really sad at this point! :(

**Reply by icyJoseph** [reply]
> yeah so, i deleted the files and added them gradually {not all}, the build succeed but as soon as i added more files, and folders in the pages folder the build starts failing.

Do you think it was a particular page/file that starts the problem?

I imagine this must be incredibly time consuming to do. 

One thing that's not so clear to me, could you try to use `NODE_OPTIONS=--max-old-space-size 2048 ` while building locally as well? Because that should limit the size of objects on your local build, and should crash locally as well.

> from this i guess we can infer that this is not a problem a next-with-less plugin.

Yeah, that should be fine now

> the last resort is to see if the increase ec2 memory helps, If that bails on me, i don't know what to do man, really sad at this point! :(

This would be a good thing to try out, however, if there's a really bad memory leak or unused objects that are not garbage collected, that might not be enough.

**Reply by the-mgi** [reply]
yeah so, i did tried with the 2048 on local and on sevrer, on local it builds fine, but on server it is bailing on me :(

**Reply by icyJoseph** [reply]
Hi,

I just thought of something. In your CircleCI config, can you find CIRCLE_NODE_TOTAL, I don't know if its in the config file, or their dashboard.

My guess is that when Next.js is building using workers, it makes the CircleCI run out of memory.

```js
const nextConfig = {
/* other config */ 
experimental: {
   cpus: Math.max(
        1,
        (Number(process.env.CIRCLE_NODE_TOTAL) ||
          (os.cpus() || { length: 1 }).length) - 1
      ),
  }
}
```

So perhaps, if you specify the `experimental.cpus` on your config, or if you figure out how to reduce the CIRCLE_NODE_TOTAL, then this would work! 

> Sometimes tools that automatically spawn workers, processes, or threads based on CPU count can inadvertently cause memory issues on CircleCI, and require manual config when running in the CI environment. Read more about how to manually limit [dynamically spawned processes here](https://support.circleci.com/hc/en-us/articles/360038192673-NodeJS-Builds-or-Test-Suites-Fail-With-ENOMEM-or-a-Timeout).

It's worth trying.

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  experimental: {
    cpus: 3,
  },
};

module.exports = nextConfig;
```

**Reply by the-mgi** [reply]
will try it for sure, but just so you know, apart from circleci, if i try to build it on the server itself after ssh, it gives me the same error

**Reply by icyJoseph** [reply]
@the-mgi did you get a change to test this out?

**Reply by the-mgi** [reply]
i tried this, but to no use. 
and in the end we dropped the idea of migrating :|
